### ISSUE TITLE
how to "retype" rdf:JSON to geo:geoJSONLiteral?

### ISSUE THREAD

**Issue by VladimirAlexiev** [w3c/json-ld-syntax#425]
https://w3c.github.io/json-ld-syntax/#json-literals shows how a piece of the JSON can be captured as datatyped literal `"..."^^rdf:JSON`

@situx @nicholascar @mathib @dr-shorthair 
https://github.com/opengeospatial/ogc-geosparql/issues/1 discusses the new GeoSPARQL 1.1 [geo:geoJSONLiteral](https://opengeospatial.github.io/ogc-geosparql/geosparql11/spec.html#_rdfs_datatype_geogeojsonliteral). According to the standard, such geometry serialization must both be connected by a specific property, and carry a specific datatype:
```ttl
<geometry> geo:asGeoJSONLiteral "..."^^geo:geoJSONLiteral
```

@gkellogg Is it possible to somehow  "retype" a `rdf:JSON` literal to a specific datatype `geo:geoJSONLiteral`?
It's easy enough to do this post-factum (eg with a SPARQL update), but seems unsatisfactory...

Perhaps something like this can be added:
```js
"@context": {
    "@version": 1.1,
    "geo": "http://www.opengis.net/ont/geosparql#",
    "myGeom": {"@id": "geo:asGeoJSONLiteral", "@type": "@json", "@json": "geo:geoJSONLiteral"}
}
```
Notice the last key: this is similar to `"@nest": "labels"` in https://w3c.github.io/json-ld-syntax/#example-defining-property-nesting

**Comment by TallTed** [comment]
It seems dangerous to retype *all* encountered `rdf:JSON` to `geo:geoJSONLiteral`, except within some specific dataset or data interaction.

I'm struggling to think of why you might need this to have happened, at any time other than during a query, at which point SPARQL/GeoSPARQL always lets you coerce/cast the datatype you want (assuming it's a "natural" transformation, which would be the case if `geo:geoJSONLiteral` is a subtype of `rdf:JSON`, as it appears)...

Would that not be sufficient?

**Comment by gkellogg** [comment]
This issue was discussed in a [meeting](https://json-ld.org/minutes/2024-03-06/#44).

<blockquote>
<div onmouseout="$('#link-44').hide()" onmouseover="$('#link-44').show()" id="44" class="information">Issue w3c/json-ld-syntax#425 <a id="link-44" style="display:none;" href="#44">✪</a></div>
<div onmouseout="$('#link-45').hide()" onmouseover="$('#link-45').show()" id="45" class="information"><a href="https://github.com/w3c/json-ld-syntax/issues/425">https://github.com/w3c/json-ld-syntax/issues/425</a> -&gt; Issue 425 how to &quot;retype&quot; rdf:JSON to geo:geoJSONLiteral? (by VladimirAlexiev) <a id="link-45" style="display:none;" href="#45">✪</a></div>
<div onmouseout="$('#link-46').hide()" onmouseover="$('#link-46').show()" id="46" class="comment"><span class="name">Niklas Lindstr&#246;m</span>:  No strong feelings about this, but my reaction is that I&apos;ve considered something like this and could be important to do, but not in JSON-LD. <a id="link-46" style="display:none;" href="#46">✪</a></div>
<div class="comment-continuation">... Whatever we might come up with could have some repercussions on how to deal with literals in RDF.</div>
<div onmouseout="$('#link-47').hide()" onmouseover="$('#link-47').show()" id="47" class="comment"><span class="name">Gregg Kellogg</span>:  I think this is narrowly defined on specifying the datatype of a JSON literal. <a id="link-47" style="display:none;" href="#47">✪</a></div>
<div onmouseout="$('#link-48').hide()" onmouseover="$('#link-48').show()" id="48" class="comment"><span class="name">Niklas Lindstr&#246;m</span>:  If properties are intrinsic in the data-space of the datatype, you could think of what it entails. <a id="link-48" style="display:none;" href="#48">✪</a></div>
<div class="comment-continuation">... It reminds me of the direction thing for language literals.</div>
<div class="comment-continuation">... My hope would be that it could serve a general purpose for RDF usage.</div>
<div onmouseout="$('#link-49').hide()" onmouseover="$('#link-49').show()" id="49" class="comment"><span class="name">David I. Lehn</span>:  When I looked at this, I saw it in the RDF domain. Could it be done at the application level? <a id="link-49" style="display:none;" href="#49">✪</a></div>
<div class="comment-continuation">... Is it a general solution?</div>
<div onmouseout="$('#link-50').hide()" onmouseover="$('#link-50').show()" id="50" class="comment"><span class="name">Pierre-Antoine Champin</span>:  What you said about the asymmetry is interesting. There are corner cases where JSON-LD fails to do proper coercion. <a id="link-50" style="display:none;" href="#50">✪</a></div>
<div onmouseout="$('#link-51').hide()" onmouseover="$('#link-51').show()" id="51" class="comment"><span class="name">Pierre-Antoine Champin</span>: &quot;P&quot;: 3.14 <a id="link-51" style="display:none;" href="#51">✪</a></div>
<div class="comment-continuation">... For example JSON numbers.</div>
<div onmouseout="$('#link-52').hide()" onmouseover="$('#link-52').show()" id="52" class="comment"><span class="name">Pierre-Antoine Champin</span>: &quot;314E-2&quot;^^xsd:decimal <a id="link-52" style="display:none;" href="#52">✪</a></div>
<div class="comment-continuation">... Things are not as smooth as you describe.</div>
<div class="comment-continuation">... The only robust way to get a literal of any type is to use strings. You get into trouble using numbers, and you could using JSON values, as well.</div>
<div class="comment-continuation">... Not sure it&apos;s a bug; I&apos;d lean towards application level.</div>
<div onmouseout="$('#link-53').hide()" onmouseover="$('#link-53').show()" id="53" class="comment"><span class="name">Benjamin Young</span>:  I think this is severe scope creep; if we add geoJSONliteral, it gets nuts. <a id="link-53" style="display:none;" href="#53">✪</a></div>
<div class="comment-continuation">... We could get data peppered with all sorts of different kinds of literals. Ultimately, if it&apos;s a string, you can parse it into JSON.</div>
<div class="comment-continuation">... Better stated using properties within the resulting graph, than by expecting that the content of the literal has some additional meaning.</div>
<div onmouseout="$('#link-54').hide()" onmouseover="$('#link-54').show()" id="54" class="comment"><span class="name">Gregg Kellogg</span>:  TAG thinks polyglot formats are an anti-pattern. <a id="link-54" style="display:none;" href="#54">✪</a></div>
<div onmouseout="$('#link-55').hide()" onmouseover="$('#link-55').show()" id="55" class="comment"><span class="name">Benjamin Young</span>:  This could lead to a proliferation of datatypes that require special knowledge to understand. <a id="link-55" style="display:none;" href="#55">✪</a></div>
<div onmouseout="$('#link-56').hide()" onmouseover="$('#link-56').show()" id="56" class="comment"><span class="name">Gregg Kellogg</span>:  Propose closing. <a id="link-56" style="display:none;" href="#56">✪</a></div>
<div onmouseout="$('#link-57').hide()" onmouseover="$('#link-57').show()" id="57" class="comment"><span class="name">David I. Lehn</span>:  This may beyond the scope of the group, but it feels like structured suffixes of media-types. <a id="link-57" style="display:none;" href="#57">✪</a></div>
<div class="comment-continuation">... Not sure how to represent that in RDF.</div>
<div onmouseout="$('#link-58').hide()" onmouseover="$('#link-58').show()" id="58" class="comment"><span class="name">Pierre-Antoine Champin</span>: Geo:geoJSONLiteral rdfs:subClassOf rdf:JSON <a id="link-58" style="display:none;" href="#58">✪</a></div>
<div onmouseout="$('#link-59').hide()" onmouseover="$('#link-59').show()" id="59" class="comment"><span class="name">Ted Thibodeau Jr.</span>:  It would be a sub-property. <a id="link-59" style="display:none;" href="#59">✪</a></div>
<div onmouseout="$('#link-60').hide()" onmouseover="$('#link-60').show()" id="60" class="comment"><span class="name">Ted Thibodeau Jr.</span>: &quot;Blah&quot;^^<a href="https://www.w3.org/ns/iana/media-types/application/geo+json">https://www.w3.org/ns/iana/media-types/application/geo+json</a> <a id="link-60" style="display:none;" href="#60">✪</a></div>
<div onmouseout="$('#link-61').hide()" onmouseover="$('#link-61').show()" id="61" class="comment"><span class="name">Gregg Kellogg</span>:  You would need to parse a string value and parse to JSON to work with it in the application layer. <a id="link-61" style="display:none;" href="#61">✪</a></div>
<div onmouseout="$('#link-62').hide()" onmouseover="$('#link-62').show()" id="62" class="information">Q <a id="link-62" style="display:none;" href="#62">✪</a></div>
<div onmouseout="$('#link-63').hide()" onmouseover="$('#link-63').show()" id="63" class="comment"><span class="name">Niklas Lindstr&#246;m</span>:  I agree with the use cases and pragmatic within the JSON-LD context, but it opens up several strange things. <a id="link-63" style="display:none;" href="#63">✪</a></div>
<div class="comment-continuation">... The real solution is to use RDF to describe things with properties.</div>
<div onmouseout="$('#link-64').hide()" onmouseover="$('#link-64').show()" id="64" class="comment"><span class="name">Ted Thibodeau Jr.</span>: Or probably better ... &quot;blah&quot;^^<a href="https://www.iana.org/assignments/media-types/application/geo+json">https://www.iana.org/assignments/media-types/application/geo+json</a> <a id="link-64" style="display:none;" href="#64">✪</a></div>
<div class="comment-continuation">... In practice we use structured values. Theory and practice often collide.</div>
<div class="comment-continuation">... Not convinced to close it, but can see why we might want to defer to the application layer.</div>
<div onmouseout="$('#link-65').hide()" onmouseover="$('#link-65').show()" id="65" class="comment"><span class="name">Benjamin Young</span>:  It&apos;s a heavy topic that we should continue to discuss. I agree it should be in the application layer. <a id="link-65" style="display:none;" href="#65">✪</a></div>
<div class="comment-continuation">... The reason we did this is because of query considerations, and a query engine might be expected to introspect these values.</div>
<div class="comment-continuation">... From GeoSPARQL, you want to be able to find GeoJSON literals.</div>
<div class="comment-continuation">... This requires that the application do more work of the database.</div>
<div class="comment-continuation">... This requires vocabularies to describe the relationship between GeoJSON and RDF.</div>
<div class="comment-continuation">... What&apos;s being asked for is a way of doing magic-string labeling, which is what media-types are.</div>
<div class="comment-continuation">... A better solution would be to figure out how to bring in media-types and then we could leverage this.</div>
<div class="comment-continuation">... But, I don&apos;t just want to add new terms to do a surface solution that doesn&apos;t really solve the problem.</div>
</blockquote>

**Comment by VladimirAlexiev** [comment]
To [Benjamin Young](https://github.com/search?q=Benjamin+Young&type=users) (Github finds 23 users of that name so I don't know whom to ping):
- JSONLD already has `"@type": "@json"` and has introduced datatype `rdf:JSON`.
  - All I'm asking is to add a key `"@json"` to allow me to substitute a datatype of my choosing.
  - Just like people have registered **media types** more specific than JSON (eg `ld+json`, `geo+json` etc), 
    there is a need for **datatypes** more specific than `rdf:JSON`.
- "this is severe scope creep; if we add geoJSONliteral, it gets nuts":
  - I have not asked to add `geoJSONliteral` to JSON-LD!
- "This requires vocabularies to describe the relationship between GeoJSON and RDF":
  - There is such specification & ontology: GeoSPARQL.

To @gkellogg 
- "You would need to parse a string value and parse to JSON to work with it in the application layer"
- "This could lead to a proliferation of datatypes that require special knowledge to understand."
  - Yes, that's what GeoSPARQL specifies.   It's a decision taken in GeoSPARQL 1.0 that was released in 2012-09-10.
  - I personally think that GeoSPARQL's decision to keep geometries in opaque literals rather than breaking them up into billions of triples is great!

To @TallTed:
- Or probably better  `"blah"^^https://www.iana.org/assignments/media-types/application/geo+json` 
  - It might have been better, but GeoSPARQL 1.1 has specified `geo:geoJSONLiteral` (and geo:wktLiteral 12 years earlier)
  - Your own GeoSPARQL Benchmark uses [wktLiteral, gmlLiteral](https://github.com/OpenLinkSoftware/GeoSPARQLBenchmark/blob/master/src/main/resources/gsb_dataset/dataset.rdf#L133), so why the sudden resistance to `geoJSONLiteral`

Notes:
- Virtuoso [seems](https://github.com/openlink/virtuoso-opensource/blob/02d09389292acd98e8c4ce50719a1f597689257b/libsrc/Wi/xqf.c#L3605) to [support](https://github.com/openlink/virtuoso-opensource/blob/02d09389292acd98e8c4ce50719a1f597689257b/libsrc/Wi/rdf_core.c#L3547) `wktLiteral` but maybe not `gmlLiteral`
- Our GraphDB supports `wktLiteral, gmlLiteral` but not the new (1.1) `geoJSONLiteral, kmlLitral`

In summary: I don't think it's appropriate for JSON-LD to question other standard group's decisions on what to keep as literals and what to break down into triples.
JSON-LD already has key `@json` to emit `rdf:JSON` literals: I'm only asking for a way to attach a more specific datatype.

Cheers!

**Comment by TallTed** [comment]
@VladimirAlexiev 

> To [Benjamin Young](https://github.com/search?q=Benjamin+Young&type=users) (Github finds 23 users of that name so I don't know whom to ping):

I believe you're looking for @BigBlueHat

> To @TallTed:
>
> * Or probably better `"blah"^^https://www.iana.org/assignments/media-types/application/geo+json`
>   * It might have been better, but GeoSPARQL 1.1 has specified `geo:geoJSONLiteral` (and `geo:wktLiteral` 12 years earlier)
>   * Your own GeoSPARQL Benchmark uses [wktLiteral, gmlLiteral](https://github.com/OpenLinkSoftware/GeoSPARQLBenchmark/blob/master/src/main/resources/gsb_dataset/dataset.rdf#L133), so why the sudden resistance to `geoJSONLiteral`?

That GeoSPARQL Benchmark belongs to [OpenLink Software](https://github.com/OpenLinkSoftware), my employer; it is not "[my] own".

"Sudden resistance"? Conversation sparks thoughts, which get "voiced" in IRC. I think you're reading more into those thoughts/comments than was meant.

Also, it appears that you may have misread my "probably better" to have been saying that `https://www.iana.org/assignments/media-types/application/geo+json` would probably be better than `geo:geoJSONLiteral`, when I was saying that `https://www.iana.org/assignments/media-types/application/geo+json` would probably be better than my previously offered `https://www.w3.org/ns/iana/media-types/application/geo+json`.

Still, I see no problem with raising the possibility of `https://www.iana.org/assignments/media-types/application/geo+json` (or `https://www.w3.org/ns/iana/media-types/application/geo+json`) as a datatype, which might be defined to be equivalent to [`geo:geoJSONLiteral`](http://www.opengis.net/ont/geosparql#geoJSONLiteral).

Interestingly, [RFC7946, The GeoJSON Format](https://datatracker.ietf.org/doc/html/rfc7946), which has no "obsoleted by", does not contain the string `geoJSONLiteral`, so it's not clear to me that "GeoSPARQL 1.1 has specified `geo:geoJSONLiteral`". There are references to external documents, but these lack dereferenceable URIs, not even pointing at documents behind paywalls, so it's difficult if not impossible for new readers to learn more.

> In summary: I don't think it's appropriate for JSON-LD to question other standard group's decisions on what to keep as literals and what to break down into triples.

I don't think it's appropriate for anyone to tell me (nor the JSON-LD WG) that I (we) cannot break some literal(s) into triples. Whether that anyone chooses to consume or otherwise make use of the triples I (we) have derived from those literals is up to them.

> JSON-LD already has key `@json` to emit `rdf:JSON` literals: I'm only asking for a way to attach a more specific datatype.

I think this has already been answered. *In general*, "key `@json` [should be left] to emit `rdf:JSON` literals". In *your specific document(s)*, where you *know* that all now generic `rdf:JSON` literals are really `geo:geoJSONLiteral` literals, you can of course have your own `@context` which maps `@json` to emit `geo:geoJSONLiteral` literals. But this should not be a general/global/universal `@context` declaration! That's part of the point of `@context` — its content is *context specific*.

**Comment by VladimirAlexiev** [comment]
@TallTed GeoJSON is not obsoleted by GeoSPARQL: it is incorporated into it.
See https://docs.ogc.org/is/22-047r1/22-047r1.html#10-8-3-1-%C2%A0-rdfs-datatype-geo-geojsonliteral
(By the same token, GML is not obsoleted by GeoSPARQL: it is incorporated as another geometry serialization format.)

> I don't think it's appropriate for anyone to tell me (nor the JSON-LD WG) that I (we) cannot break some literal(s) into triples.

This approach will break established implementation practices (semantic repositories that follow the GeoSPARQL standard to handle geometries). What would be the benefit of doing that?

> you can of course have your own `@context` which maps `@json` to emit `geo:geoJSONLiteral` literals. 

How? That's exactly the possibility I'm asking for.

**Comment by TallTed** [comment]
Anyone can mint their own `@context` document to reference at the top of their own JSON-LD documents. As I understand your described wish, which is simply [type coercion](https://w3c.github.io/json-ld-syntax/#type-coercion), I believe you could put this at the start of your GeoJSON documents —

```json
"@context": {
    "geo": "http://www.opengis.net/ont/geosparql#",
    "myGeom": {"@id": "geo:asGeoJSONLiteral", "@type": "geo:geoJSONLiteral"}
}
```
— or perhaps better —
```json
"@context": {
    "myGeom": {"@id": "http://www.opengis.net/ont/geosparql#asGeoJSONLiteral", 
               "@type": "http://www.opengis.net/ont/geosparql#geoJSONLiteral"}
}
```

Note that `"@version": 1.1,` as was seen in your initial example, is not needed.

Although... I can find nothing that discusses the `geo:asGeoJSONLiteral` property you say is required to be used. `http://www.opengis.net/ont/geosparql#asGeoJSONLiteral` does not dereference; there is no `asGeoJSONLiteral` fragment in the documents returned from `http://www.opengis.net/ont/geosparql`.

You may need to remove or edit any existing `@context` declaration to remove some existing mapping(s), particularly if those existing `@context` include any `@protected` settings.

**Comment by VladimirAlexiev** [comment]
Hi @TallTed!

Please read https://w3c.github.io/json-ld-syntax/#json-literals.
You have to state `"@type": "@json"` if you want a fragment of JSON to be captured into a literal.

Let's try at the playground:  https://tinyurl.com/287gzb57
```js
{
  "@context": {
    "@version": 1.1,
    "geo": "http://www.opengis.net/ont/geosparql#",
    "asGeoJSON": {"@id": "geo:asGeoJSON", "@type": "@json"}
  },
  "asGeoJSON": [1,2,3]
}
```
returns what I want, except for the generic datatype:
```ttl
_:b0 <http://www.opengis.net/ont/geosparql#asGeoJSON> "[1,2,3]"^^<http://www.w3.org/1999/02/22-rdf-syntax-ns#JSON> .
```

If I follow your suggestion:
```js
{
  "@context": {
    "@version": 1.1,
    "geo": "http://www.opengis.net/ont/geosparql#",
    "asGeoJSON": {"@id": "geo:asGeoJSON", "@type": "geo:geoJSONLiteral"}
  },
  "asGeoJSON": [1,2,3]
}
```
then the geometry is broken into individual triples, something I don't want:
```ttl
_:b0 <http://www.opengis.net/ont/geosparql#asGeoJSON> "1"^^<http://www.opengis.net/ont/geosparql#geoJSONLiteral> .
_:b0 <http://www.opengis.net/ont/geosparql#asGeoJSON> "2"^^<http://www.opengis.net/ont/geosparql#geoJSONLiteral> .
_:b0 <http://www.opengis.net/ont/geosparql#asGeoJSON> "3"^^<http://www.opengis.net/ont/geosparql#geoJSONLiteral> .
```

If the input JSON has a **string** that looks like a JSON fragment, then sure I can coerce it to a specific datatype:
```js
{
  "@context": {
    "@version": 1.1,
    "geo": "http://www.opengis.net/ont/geosparql#",
    "asGeoJSON": {"@id": "geo:asGeoJSON", "@type": "geo:geoJSONLiteral"}
  },
  "asGeoJSON": "[1,2,3]"
}
```
But the input files that I work with (JSON-FG, CityJSON etc) are all JSON, they don't use strings looking like JSON.

Now read my original proposal: I want a new key `@json` (mimicking the existing value `@json`)
so I can retype to a specific datatype:
```js
"asGeoJSON": {"@id": "geo:asGeoJSON", "@type": "@json", "@json": "geo:geoJSONLiteral"}
```
Is it more clear now?

---

> There is no asGeoJSONLiteral fragment in the documents returned from http://www.opengis.net/ont/geosparql.

Sorry, my mistake, here's the correct prop pointing to the correct datatype:
```ttl
:asGeoJSON
	a rdf:Property, owl:DatatypeProperty ;
	rdfs:subPropertyOf :hasSerialization;
	rdfs:domain :Geometry ;
	rdfs:range :geoJSONLiteral ;
```

**Comment by BigBlueHat** [comment]
> ```js
> "asGeoJSON": {"@id": "geo:asGeoJSON", "@type": "@json", "@json": "geo:geoJSONLiteral"}
> ```

This gets close to what's now in my head after reading all the defense/feedback, though with one key change (for a future version of JSON-LD).

```jsonc
{
  "@context": {
    "asGeoJSON": {"@id": "geo:asGeoJSON", "@type": "geo:geoJSONLiteral", "@json": "object"}
  },
  "asGeoJSON": { /* raw geojson */ }
}
```
resulting in:
```turtle
_:b0 <geo:asGeoJSON> "{ /* raw geojson */ }"^^<geo:geoJSONLiteral>
```

That proposed format could potentially allow us to more clearly unconflate RDF datatypes and JSON value/format types. The proposed `@json` key could be used to let the processor know to encode the JSON object as a value for RDF processors and `@type` could then be focused solely on the RDF datatype.

Alternatively, the JSON object storage (vs. processing as more JSON-LD) could be done automatically when objects are detected--though this could result in unexpected processing output if/when errors occurred in the JSON...and undoubtedly makes the "is this bit JSON-LD or just JSON?" a harder question for authors to keep straight.

But...could result in this "just working":
```jsonc
{
  "@context": {
    "asGeoJSON": {"@id": "geo:asGeoJSON", "@type": "geo:geoJSONLiteral"}
  },
  "asGeoJSON": { /* raw geojson */ }
}
```

I'm certain there's issues with that reason that @gkellogg and others can elaborate on, but it may be interesting to explore if/when we could make datatypes more flexible...at least around JSON objects as values.

**Comment by VladimirAlexiev** [comment]
@BigBlueHat Thanks for supporting this!
I like better `"@type": "geo:geoJSONLiteral"`. 
But for the value of `@json`:
- Don't you need a key not plain string, eg like `@object`?
- Or maybe we can use a boolean: `@json: true`
- Or here's a silly suggestion `@json: @json` :-)

PS: when you take into account the prefix `"geo": "http://www.opengis.net/ont/geosparql#"`, 
this results in prefixed URLs: `geo:asGeoJSON "{ /* raw geojson */ }"^^geo:geoJSONLiteral`.

The difference is crucial: there is in fact a `geo` URI scheme, and 
`<geo:12.3,45.6>` is a valid geo point, whereas `<geo:asGeoJSON>` is not a valid URI.

**Comment by gkellogg** [comment]
This relates to work on [CDT](https://awslabs.github.io/SPARQL-CDTs/spec/latest.html) literals proposed by AWS for SPARQL. Being about to use `cdt:Map` or `cdt:List` as the datatype of such a literal would directly play into this mechanism. See https://github.com/awslabs/SPARQL-CDTs/issues/3#issuecomment-2143547910.

**Comment by VladimirAlexiev** [comment]
If the need for such "retyping" has been established, can we consider concrete syntaxes to express it? Here are some choices:
```
1. "@type": "@json", "@json": "geo:geoJSONLiteral"
2. "@type": "geo:geoJSONLiteral", "@json": true
3. "@type": "geo:geoJSONLiteral", "@json": "@json"
4. "@type": ["@json", "geo:geoJSONLiteral"]
```
@gkellogg what do you think?

**Comment by gkellogg** [comment]
We probably need to think about other use cases that could use the same pattern. Really, the issue is when `@type` is used with a keyword and the resulting datatype IRI is implicit. Is there some corollary for YAML or CBOR that is not handled by `@json`?

Another possibility would be to add `@json` as a possible value for `@container`, which would treat the value as JSON but allow another datatype to be specified using `@type`. For example:

```json
{
  "@context": {
    "asGeoJSON": {"@id": "geo:asGeoJSON", "@container": "@json", "@type": "geo:geoJSONLiteral"}
  },
  "asGeoJSON": { /* raw geojson */ }
}
```

**Comment by dr-shorthair** [comment]
@TallTed have you seen `geo:` used in practice? 
I note that the registry [1] has not been touched since the RFC was issued. 

[1] https://www.iana.org/assignments/geo-uri-parameters/geo-uri-parameters.xhtml#geo-uri-parameters-1

**Comment by TallTed** [comment]
@dr-shorthair — I was following up on https://github.com/w3c/json-ld-syntax/issues/425#issuecomment-2138882482 (which I should have cited). I have not personally seen [RFC5870's `geo:`](https://datatracker.ietf.org/doc/html/rfc5870) in use in the wild, but that means nothing; there are thousands if not millions of RDF users out there, and I don't see all their data.

**Comment by VladimirAlexiev** [comment]
@gkellogg I like `"@container": "@json"`.
As for extending into other "BLOB/CLOB" types (YAML, CBOR etc), I think this is a far future extension...
